fem/spacehierarchy.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#include "spacehierarchy.hpp"
+
+namespace mfem
+{
+
+template <typename M, typename FES>
+AbstractSpaceHierarchy<M, FES>::AbstractSpaceHierarchy(M* mesh, FES* fespace,
+                                                       bool ownM, bool ownFES)
+{
+   AddLevel(mesh, fespace, ownM, ownFES);
+}
+
+template <typename M, typename FES>
+AbstractSpaceHierarchy<M, FES>::~AbstractSpaceHierarchy()
+{
+   for (int i = meshes.Size() - 1; i >= 0; --i)
+   {
+      if (ownedFES[i])
+      {
+         delete fespaces[i];
+      }
+      if (ownedMeshes[i])
+      {
+         delete meshes[i];
+      }
+   }
+
+   fespaces.DeleteAll();
+   meshes.DeleteAll();
+}
+
+template <typename M, typename FES>
+unsigned AbstractSpaceHierarchy<M, FES>::GetNumLevels() const
+{
+   return meshes.Size();
+}
+
+template <typename M, typename FES>
+unsigned AbstractSpaceHierarchy<M, FES>::GetFinestLevelIndex() const
+{
+   return GetNumLevels() - 1;
+}
+
+template <typename M, typename FES>
+void AbstractSpaceHierarchy<M, FES>::AddLevel(M* mesh, FES* fespace, bool ownM,
+                                              bool ownFES)
+{
+   meshes.Append(mesh);
+   fespaces.Append(fespace);
+   ownedMeshes.Append(ownM);
+   ownedFES.Append(ownFES);
+}
+
+template <typename M, typename FES>
+void AbstractSpaceHierarchy<M, FES>::AddUniformlyRefinedLevel(int dim,
+                                                              int ordering)
+{
+   M* mesh = new M(*meshes[GetFinestLevelIndex()]);
+   mesh->UniformRefinement();
+   FES* coarseFEspace = fespaces[GetFinestLevelIndex()];
+   FES* fineFEspace = new FES(mesh, coarseFEspace->FEColl(), dim, ordering);
+   AddLevel(mesh, fineFEspace, true, true);
+}
+
+template <typename M, typename FES>
+void AbstractSpaceHierarchy<M, FES>::AddOrderRefinedLevel(
+    FiniteElementCollection* fec, int dim, int ordering)
+{
+   M* mesh = &GetFinestMesh();
+   FES* newFEspace = new FES(mesh, fec, dim, ordering);
+   AddLevel(mesh, newFEspace, false, true);
+}
+
+template <typename M, typename FES>
+const M& AbstractSpaceHierarchy<M, FES>::GetMeshAtLevel(unsigned level) const
+{
+   MFEM_ASSERT(level < meshes.Size(), "Mesh at given level does not exist.");
+   return *meshes[level];
+}
+
+template <typename M, typename FES>
+M& AbstractSpaceHierarchy<M, FES>::GetMeshAtLevel(unsigned level)
+{
+   return const_cast<M&>(
+       const_cast<const AbstractSpaceHierarchy*>(this)->GetMeshAtLevel(level));
```

Comment:
I think it would be nicer to avoid using `const_cast` here

Reply:
You are right, it is ugly and unnecessary here. I changed this while rewriting the SpaceHierarchy as discussed in the other comment.